### PR TITLE
Added detection of custom FETCH and UID FETCH commands for IMAP

### PR DESCRIPTION
**Pull request by MonkeybreadSoftware**
Added detection of custom FETCH and UID FETCH commands, so the output is correctly (and not redirected to debug log).

### CODE REVIEW

File: lib/imap.c [hunk 1]
```diff
@@ -660,9 +660,27 @@ static CURLcode imap_perform_list(struct Curl_easy *data)
...
+        state(data, IMAP_FETCH);
+      }
+
+      /* e.g. "UID FETCH 107 BODY.PEEK[]" */
+
+      if(imap->custom_params)
+      if(strncasecompare(imap->custom, "UID", 3) == 1)
+      if(strncasecompare(imap->custom_params, " FETCH", 6) == 1) {
+        state(data, IMAP_FETCH);
+      }
```
Comment (bagder) [inline lib/imap.c:680]: 1.This code doesn't use the proper indenting, two spaces for new levels.
2.I don't think it should change state if it immediately again changes to something else. Change only once.

Comment (MonkeybreadSoftware) [reply lib/imap.c:682]: I added spaces for intending.
I think setting IMAP_FETCH twice should not be a problem.
it is much easier code to do it this way than to create one huge if to handle both cases together.

But feel free to adjust if needed.

Comment (bagder) [reply lib/imap.c:682]: The problem is primarily that it makes it harder to understand what's going on when you look at state transitions in a verbose log but also that it isn't actually changing the state if you immediately change it again so it's sort of a lie.

If we keep state changes to *actual* state changes, we keep the ability to at some point add state init logic there, the way we've done in some of our other state machines.

Comment (MonkeybreadSoftware) [reply lib/imap.c:682]: We could of course add "return result; " after changing state to fetch.
And then do the IMAP_LIST if the compares didn't set it to fetch.

**Comment by MonkeybreadSoftware** [comment]
I added a test, which seems to work here.
I hope the ID I picked is still available.

File: lib/imap.c (hunk 1 above)
Comment (MonkeybreadSoftware) [reply lib/imap.c:688]: I committed a new suggestion for the implementation with an isFetch flag, which only changes state once now.
Comments?

**Comment by MonkeybreadSoftware** [comment]
What is still needed for this to proceed and get added to one of the next releases?

**Comment by sawoods** [comment]
For reference #536, #1150 and [Issue with IMAP UID FETCH](https://curl.se/mail/lib-2014-04/0069.html)

**Comment by MonkeybreadSoftware** [comment]
Could we get this added for the current feature window?
I'd really like to remove all the work arounds, where we take the email content from debug messages.

**Comment by Anthropos888** [comment]
I also hope that this feature gets added. The UID FETCH response isn't complete so I have to get the complete response from the debug message.

**Comment by bagder** [comment]
I restarted the appveyor builds to see if the errors remain.

**Comment by bagder** [comment]
@MonkeybreadSoftware are you still looking to get this change merged? It merge conflicts now.

(We clearly lack someone with IMAP knowledge and will to push and take responsibility for IMAP changes in the project, which often makes them fall through the cracks and get forgotten.)

**Comment by MonkeybreadSoftware** [comment]
Yes, I'd love to have this merged as then I don't need to patch it into myself every release.

**Comment by MonkeybreadSoftware** [comment]
The conflict was only the test case number, so I moved 1662 to 1665 to get a free one.

**Pull request closed without merging**